### ISSUE TITLE
Fast query of NCBI's dbSNP?

### ISSUE THREAD

**Issue by privefl** [ropensci/rsnps#31]
I have a vector of SNP names of length ~530,000 (which corresponds to an Affymetrix 500K chip).

Using `NCBI_snp_query` on the first 500 SNPs takes ~70 seconds which would require ~20 hours to query all SNPs.
Surprisingly, `NCBI_snp_query2` takes only ~20 seconds, yet querying more information (a slot `data`, that I'm not interested in).

I wonder
- if there is a way to get a faster query by explicitly saying that my SNPs are **ordered** (by chromosome and physical position),
- if there is a way to not querying the slot `data` with `NCBI_snp_query2` with the hope of making the query faster,
- if there is a way to query infos on all SNPs that are known to be part of a particular GeneChip (such as Affymetrix 500K).

**Comment by haozhu233** [comment]
Hi, @privefl, as documented in the last few rows of the `NCBI_snp_query` [#L207](https://github.com/ropensci/rsnps/blob/8e1b81e61e0d4e2e412b2f060dfe54d407ed225e/R/NCBI_snp_query.R#L207), the main short board that slows down the `NCBI_snp_query` is the query rate limitation they put on NCBI. Similarly, the reason why `NCBI_snp_query2` is a bit faster is that the system sleep time is 0.33 there. As a result, I kind of feel doubtful if your request can be done without contacting the NCBI team. @sckott, do you have any thoughts on that?

**Comment by sckott** [comment]
the internals of the two functions are slightly different, using different data formats from NCBI, thus parsing is slightly different depending on the data format. Right @haozhu233 the sleep times btw requests are diff, so that adds to difference as well. 

@privefl 

For your first 2 questions, I don't know. I could ask the NCBI people. 

> if there is a way to not querying the slot data with NCBI_snp_query2 with the hope of making the query faster

What do you mean by slot data?

**Comment by privefl** [comment]
```
names(NCBI_snp_query2(snp))
[1] "summary" "data"   
```

**Comment by sckott** [comment]
@privefl so you don't want the data in `summary`?

**Comment by privefl** [comment]
@sckott I only want the data in `summary`, which is 30 times as small as all data.

**Comment by sckott** [comment]
okay, thanks for clarification

**Comment by sckott** [comment]
problem is that we do some parsing of the returned data blob, that gives the stuff in `data` - and then from that bit we get the `summary` part, So we still need to parse the data even if we don't give `data` back to the user, meaning it's no faster. 

There may be some other way to get the data, will have to keep looking for different data return types

**Comment by sckott** [comment]
@privefl sorry for the very long delay on this. just put the function `ncbi_snp_summary()` in the package - reinstall try it out: `remotes::install_github("ropensci/rsnps")`

**Comment by privefl** [comment]
@sckott 
From what I remember, I want to get the names of the genes associated which each SNP.
With `ncbi_snp_summary()` I get no gene for the first 200 SNPs. Maybe this means that these SNPs are in non-coding regions? 
Not sure. Thanks anyway.

**Comment by sckott** [comment]
can you show me an example?

**Comment by privefl** [comment]
```
ID <- c("rs3934834", "rs3737728", "rs6687776", "rs9651273", "rs4970405", 
"rs12726255", "rs2298217", "rs4970362", "rs9660710", "rs4970420", 
"rs1320565", "rs11260549", "rs9729550", "rs11721", "rs2887286", 
"rs3813199", "rs3766186", "rs7515488", "rs715643", "rs6675798", 
"rs11804831", "rs6685064", "rs2649588", "rs819980", "rs3766178", 
"rs3128342", "rs880051", "rs2296716", "rs6603793", "rs7531583", 
"rs6681938", "rs4648592", "rs7525092", "rs2474460", "rs3820011", 
"rs2803291", "rs12755035", "rs884080", "rs908742", "rs3107151", 
"rs3128291", "rs3052", "rs262688", "rs2460002", "rs6665593", 
"rs7512482", "rs2460000", "rs260513", "rs11588312", "rs7553178", 
"rs10910047", "rs2017143", "rs903916", "rs1809822", "rs2279702", 
"rs2643904", "rs2843157", "rs2645065", "rs903904", "rs2843143", 
"rs2055204", "rs2840528", "rs903914", "rs7545940", "rs7519807", 
"rs2843160", "rs2643901", "rs2843127", "rs4648633", "rs3736330", 
"rs2840532", "rs3001336", "rs1129333", "rs2494428", "rs4531246", 
"rs10910078", "rs2494626", "rs13376356", "rs11588930", "rs2477703", 
"rs3762444", "rs7535528", "rs2236395", "rs2246732", "rs2494606", 
"rs2234167", "rs6667605", "rs3748816", "rs3890745", "rs4648356", 
"rs10909880", "rs4609377", "rs4648481", "rs12409348", "rs897615", 
"rs10797342", "rs2045333", "rs3100865", "rs1456465", "rs1869970", 
"rs897634", "rs2045332", "rs1473386", "rs2925499", "rs10909845", 
"rs878201", "rs2485945", "rs12044030", "rs12046158", "rs7533362", 
"rs1572657", "rs10909852", "rs2027262", "rs3795263", "rs2142569", 
"rs10489588", "rs1109251", "rs2376821", "rs2297828", "rs2297829", 
"rs2993486", "rs1569419", "rs926246", "rs926244", "rs2817168", 
"rs2993493", "rs1890336", "rs4648453", "rs2817178", "rs2235176", 
"rs10797380", "rs7538096", "rs2817185", "rs731031", "rs1798246", 
"rs2651899", "rs2651936", "rs2817134", "rs2651929", "rs12135062", 
"rs10752733", "rs10737190", "rs7539775", "rs10492941", "rs10909901", 
"rs2817146", "rs12124147", "rs1553288", "rs2455118", "rs4648464", 
"rs3002686", "rs2455119", "rs10492940", "rs16823802", "rs10909918", 
"rs4648377", "rs2455144", "rs6670328", "rs2500278", "rs12409345", 
"rs2500277", "rs2483266", "rs1108600", "rs2500271", "rs2483274", 
"rs4494089", "rs6683273", "rs2483286", "rs2483289", "rs2483256", 
"rs946758", "rs2500286", "rs12135987", "rs1333189", "rs1538693", 
"rs868688", "rs10492939", "rs10492938", "rs17399998", "rs2493278", 
"rs871822", "rs6424074", "rs870124", "rs2493292", "rs2493291", 
"rs2493285", "rs1984069", "rs870171", "rs2493272", "rs2483250", 
"rs867814", "rs2487670", "rs2487680", "rs2185639", "rs2493314", 
"rs4648505", "rs2821040", "rs947344", "rs11585362", "rs12093117"
)

infos <- rsnps::ncbi_snp_summary(ID)
infos$gene
```

**Comment by sckott** [comment]
made a change, reinstall to get it

there seems to be what I think are gene names in the `docsum` field. tried to extract that into a field called `gene2`

```r
unique(infos$gene2)
#>  [1] "LOC105378948:105378948,RNF223:401934"
#>  [2] "C1orf159:54991"
#>  [3] NA
#>  [4] "TTLL10:254173"
#>  [5] "SDF4:51150,TNFRSF4:7293"
#>  [6] "SDF4:51150"
#>  [7] "UBE2J2:118424"
#>  [8] "LOC101928895:101928895"
#>  [9] "ATAD3B:83858"
#> [10] "SSU72:29101"
#> [11] "NADK:65220"
#> [12] "GNB1:2782"
#> [13] "CALML6:163688"
#> [14] "CFAP74:85452,LOC107984872:107984872"
#> [15] "CFAP74:85452"
#> [16] "PRKCZ:5590"
#> [17] "LOC100506504:100506504,"
#> [18] "FAAP20:199990"
#> [19] "FAAP20:199990,LOC105378592:105378592"
#> [20] "LOC105378592:105378592"
#> [21] "SKI:6497"
#> [22] "MORN1:79906"
#> [23] "LOC100129534:100129534,MORN1:79906"
#> [24] "RER1:11079"
#> [25] "PEX10:5192"
#> [26] "PLCH2:9651"
#> [27] "PANK4:55229"
#> [28] "TNFRSF14:8764"
#> [29] "LOC100996583:100996583"
#> [30] "MMEL1:79258"
#> [31] "ACTRT2:140625"
#> [32] "LINC00982:440556,PRDM16:63976"
#> [33] "PRDM16:63976"
#> [34] "LOC105378606:105378606,PRDM16:63976"
#> [35] "LOC107984909:107984909,PRDM16:63976"
#> [36] "LOC105378605:105378605,PRDM16:63976"
#> [37] "LOC105378604:105378604,PRDM16:63976"
#> [38] "ARHGEF16:27237"
#> [39] "MEGF6:1953"
```

**Comment by privefl** [comment]
Looks great!

Seems like I can only request them 400 by 400 (which is totally fine).
Seems like I can request them in parallel (for example 5000 of them takes 45 seconds with 1 core, 25 sec with 2 cores and 20 sec with 3 cores). Is it "risky" to make too many requests at once?

**Comment by privefl** [comment]
I was able to query 280,000 SNPs in less than 20 minutes using 3 cores (out of 4).

Thanks!

**Comment by sckott** [comment]
> I was able to query 280,000 SNPs in less than 20 minutes using 3 cores (out of 4).

is that fast enough for you?  

> Is it "risky" to make too many requests at once?

NCBI does have rate limits. https://www.ncbi.nlm.nih.gov/books/NBK25497/ They do allow higher rate limits when you use an API key. We should add support for that. - see new issue #58

**Comment by privefl** [comment]
Yeah it seems a reasonable time.
My second question was: if I try to be too greddy with computers having many cores,
does NCBI will kick me out eventually? (for a certain amount of time, like one day)

**Comment by sckott** [comment]
I don't know the answer to the second question. Either they have the capability to block IP addresses or not, and I don't know if they do. Let me know if you run into it. Though you may not get a clear error message saying so if it does happen

**Comment by privefl** [comment]
Now, I got this error coming from nowhere: Bad Gateway (HTTP 502). Have you any idea where it comes from?
